infrastructure/devicepathresolver/iscsi_device_path_resolver.go

```diff
@@ -0,0 +1,184 @@
+package devicepathresolver
+
+import (
+	"fmt"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"time"
+
+	boshopeniscsi "github.com/cloudfoundry/bosh-agent/platform/openiscsi"
+	boshsettings "github.com/cloudfoundry/bosh-agent/settings"
+	boshdirs "github.com/cloudfoundry/bosh-agent/settings/directories"
+	bosherr "github.com/cloudfoundry/bosh-utils/errors"
+	boshlog "github.com/cloudfoundry/bosh-utils/logger"
+	boshsys "github.com/cloudfoundry/bosh-utils/system"
+)
+
+// iscsiDevicePathResolver resolves device path by performing Open-iscsi discovery
+type iscsiDevicePathResolver struct {
+	diskWaitTimeout time.Duration
+	runner          boshsys.CmdRunner
+	openiscsi       boshopeniscsi.OpenIscsi
+	fs              boshsys.FileSystem
+	dirProvider     boshdirs.Provider
+	logTag          string
+	logger          boshlog.Logger
+}
+
+func NewIscsiDevicePathResolver(
+	diskWaitTimeout time.Duration,
+	runner boshsys.CmdRunner,
+	openiscsi boshopeniscsi.OpenIscsi,
+	fs boshsys.FileSystem,
+	dirProvider boshdirs.Provider,
+	logger boshlog.Logger,
+) DevicePathResolver {
+	return iscsiDevicePathResolver{
+		diskWaitTimeout: diskWaitTimeout,
+		runner:          runner,
+		openiscsi:       openiscsi,
+		fs:              fs,
+		dirProvider:     dirProvider,
+		logTag:          "iscsiResolver",
+		logger:          logger,
+	}
+}
+
+func (ispr iscsiDevicePathResolver) GetRealDevicePath(diskSettings boshsettings.DiskSettings) (string, bool, error) {
```

Comment:
Is there a way to break this function some more... it's long, hard to read, and likely difficult to test.

Breaking it down might help.

Reply:
Seems like it does all of these things which could be broken down into smaller functions... Just a suggestion about how to break down. But this needs simplifications so it can be tested.

```
... GetRealDevicePath( ...) {
  // Settings check
  ...
  // Run command to discover

  // Deal with no device found case
  ...
  
  // Deal with device found: setup, discover, login
  ...
  
  // Wait for device to appear
  ...
}
```

Comments need not be included in code if you chose method names wisely.